base/linalg/rowvector.jl

```diff
@@ -0,0 +1,183 @@
+immutable RowVector{T,V<:AbstractVector} <: AbstractMatrix{T}
+    vec::V
+    function RowVector(v::V)
+        check_types(T,v)
+        new(v)
+    end
+end
+
+@inline check_types{T1,T2}(::Type{T1},::AbstractVector{T2}) = check_types(T1, T2)
+@pure check_types{T1,T2}(::Type{T1},::Type{T2}) = T1 === transpose_type(T2) ? nothing : error("Element type mismatch. Tried to create a `RowVector{$T1}` from an `AbstractVector{$T2}`")
+
+# The element type is transformed as transpose is recursive
+@inline transpose_type{T}(::Type{T}) = promote_op(transpose, T)
+
+# Constructors that take a vector
+@inline RowVector{T}(vec::AbstractVector{T}) = RowVector{transpose_type(T),typeof(vec)}(vec)
+@inline (::Type{RowVector{T}}){T}(vec::AbstractVector{T}) = RowVector{T,typeof(vec)}(vec)
+
+# Constructors that take a size and default to Array
+@inline (::Type{RowVector{T}}){T}(n::Int) = RowVector{T}(Vector{transpose_type(T)}(n))
+@inline (::Type{RowVector{T}}){T}(n1::Int, n2::Int) = n1 == 1 ? RowVector{T}(Vector{transpose_type(T)}(n2)) : error("RowVector expects 1×N size, got ($n1,$n2)")
+@inline (::Type{RowVector{T}}){T}(n::Tuple{Int}) = RowVector{T}(Vector{transpose_type(T)}(n[1]))
+@inline (::Type{RowVector{T}}){T}(n::Tuple{Int,Int}) = n[1] == 1 ? RowVector{T}(Vector{transpose_type(T)}(n[2])) : error("RowVector expects 1×N size, got $n")
+
+# similar()
+@inline similar(rowvec::RowVector) = RowVector(similar(rowvec.vec))
+@inline similar{T}(rowvec::RowVector, ::Type{T}) = RowVector(similar(rowvec.vec, transpose_type(T)))
+# There is no resizing similar() because it would be ambiguous if the result were a Matrix or a RowVector
+
+# Basic methods
+@inline transpose(vec::AbstractVector) = RowVector(vec)
+@inline ctranspose{T}(vec::AbstractVector{T}) = RowVector(conj(vec))
+@inline ctranspose{T<:Real}(vec::AbstractVector{T}) = RowVector(vec)
+
+@inline transpose(rowvec::RowVector) = rowvec.vec
+@inline ctranspose{T}(rowvec::RowVector{T}) = conj(rowvec.vec)
+@inline ctranspose{T<:Real}(rowvec::RowVector{T}) = rowvec.vec
+
+# Some overloads from Base
+@inline transpose(r::Range) = RowVector(r)
+@inline ctranspose(r::Range) = RowVector(conj(r)) # is there such a thing as a complex range?
```

Comment:
I like the approach in #6837 (which doesn't make a copy for `ctranspose`) better.